protocols/relay/src/v2/protocol/outbound_hop.rs

```diff
@@ -0,0 +1,221 @@
+use libp2p_swarm::NegotiatedSubstream;
+use prost::Message;
+use std::convert::TryFrom;
+use std::io::Cursor;
+use std::iter;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use thiserror::Error;
+use unsigned_varint::codec::UviBytes;
+
+pub enum Upgrade {
+    Reserve,
+    Connect { dst_peer_id: PeerId },
+}
+
+impl upgrade::UpgradeInfo for Upgrade {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(HOP_PROTOCOL_NAME)
+    }
+}
+
+impl upgrade::OutboundUpgrade<NegotiatedSubstream> for Upgrade {
+    type Output = Output;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(self, substream: NegotiatedSubstream, _: Self::Info) -> Self::Future {
+        let msg = match self {
+            Upgrade::Reserve => HopMessage {
+                r#type: hop_message::Type::Reserve.into(),
+                peer: None,
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+            Upgrade::Connect { dst_peer_id } => HopMessage {
+                r#type: hop_message::Type::Connect.into(),
+                peer: Some(Peer {
+                    id: dst_peer_id.to_bytes(),
+                    addrs: vec![],
+                }),
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+        };
+
+        let mut encoded_msg = Vec::new();
+        msg.encode(&mut encoded_msg)
+            .expect("Vec to have sufficient capacity.");
+
+        let mut codec = UviBytes::default();
+        codec.set_max_len(MAX_MESSAGE_SIZE);
+        let mut substream = Framed::new(substream, codec);
+
+        async move {
+            substream.send(std::io::Cursor::new(encoded_msg)).await?;
+            let msg: bytes::BytesMut = substream
+                .next()
+                .await
+                .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::UnexpectedEof, ""))??;
+
+            let HopMessage {
+                r#type,
+                peer: _,
+                reservation,
+                limit: _,
+                status,
+            } = HopMessage::decode(Cursor::new(msg))?;
+
+            let r#type = hop_message::Type::from_i32(r#type).ok_or(UpgradeError::ParseTypeField)?;
+            match r#type {
+                hop_message::Type::Connect => return Err(UpgradeError::UnexpectedTypeConnect),
+                hop_message::Type::Reserve => return Err(UpgradeError::UnexpectedTypeReserve),
+                hop_message::Type::Status => {}
+            }
+
+            let status = Status::from_i32(status.ok_or(UpgradeError::MissingStatusField)?)
+                .ok_or(UpgradeError::ParseStatusField)?;
+            match status {
+                Status::Ok => {}
+                s => return Err(UpgradeError::UnexpectedStatus(s)),
```

Comment:
I think `UnexpectedStatus` is a bit misleading here. Imo we have to differentiate here between actually unexpected states (e..g in this case `ConnectionFailed`, `NoReservation`) that validate protocol, and states that are perfectly valid, like `ReservationRefused` and `ResourceLimitExceeded`. This differentiation is already implemented in the `client::handler`, but i think it would be cleaner to already do it here, e.g. 
```rust
pub enum UpgradeError {
  ... 
  UnexpectedStatus, // what the unexpected status was is acutally not really relevant
  ReservationFailed(ReservationDenied)
}
pub enum ReservationDenied {
  Refused,
  LimitExceeded
}
```
The `ReservationDenied` enum could then also be reused on behaviour events on the higher layer, 
in `relay::Event::ReservationReqDenied` and `client::Event::ReservationReqAccepted`.

In general, I think the `UpgradeError`s throughout all of the 4 protocols should be a bit cleaner, and reduced to the for the user actually relevant information. 
For example `outbound_hop::UpgradeError`:
```rust 
pub enum UpgradeError {
    Decode(prost::DecodeError,),
    Io(std::io::Error),
    MissingStatusField,
    MissingReservationField,
    NoAddressesInReservation,
    InvalidReservationExpiration,
    InvalidReservationAddrs,
    ParseTypeField,
    UnexpectedTypeConnect,
    UnexpectedTypeReserve,
    ParseStatusField,
    UnexpectedStatus(Status),
}
```
Each of these 12 different variant can in practise be categorized into one of the following 3 errors, that are actually relevant:
1. Reading/ Writing from/ to I/O failed
2. What we read from I/O is not what we expected to read -> Protocol Violation
3. (Actually not an error really): Our reservation was rejected.  

In case Nr2. it doesn't really matter if the decoding directly failed or a wrong status was present/ an expected field wasn't present/ etc.; for the user imo it doesn't make a difference. Therefore we can reduce the 12 variants into these 3:
```
pub enum UpgradeError {
    DecodeFailed,
    Io(std::io::Error),
     ReservationFailed(ReservationDenied)
}
```
For the first two variant the handler will then create a fatal error, so maybe we could clean that even more:
```rust 
pub enum UpgradeError {
    Fatal(FatalUpgradeError),
     ReservationFailed(ReservationDenied)
}

pub enum FatalUpgradeError {
    DecodeFailed,
    Io(std::io::Error),
}
```
and then in the handler: 
```rust
fn inject_dial_upgrade_error(..) {
  match open_info {
     OutboundOpenInfo::Connect {..} => {
        match error {
          ProtocolsHandlerUpgrErr::Upgrade(upgrade::UpgradeError::Apply(error)) => {
              match error {
                  outbound_hop::UpgradeErrorFatal(error) => {
                      self.pending_error = Some(ProtocolsHandlerUpgrErr::Upgrade(
                          upgrade::UpgradeError::Apply(EitherError::B(error)),
                      ));
                  }
                  outbound_hop::ReservationFailed(_) => {
                          // While useless for reaching this particular destination, the
                          // connection to the relay might still proof helpful for other
                          // destinations. Thus do not terminate the connection.
                  }
```
What do you think about this? (Also referencing a discussion about error types we had in our team: https://github.com/iotaledger/identity.rs/discussions/460#discussioncomment-1558171)

Reply:
What do you think of https://github.com/libp2p/rust-libp2p/commit/2f5b8e6393286773abe66423b99583b26e2ecb49? Should I apply the same patterns to the other 3 `UpgradeError`s.

> I think UnexpectedStatus is a bit misleading here. Imo we have to differentiate here between actually unexpected states (e..g in this case ConnectionFailed, NoReservation) that violate the protocol, and states that are perfectly valid, like ReservationRefused and ResourceLimitExceeded. 

Done in the commit above.

> The ReservationDenied enum could then also be reused on behaviour events on the higher layer,
in relay::Event::ReservationReqDenied and client::Event::ReservationReqDenied.

Not sure I follow. In my proposal (see commit above) would one bubble up `ReservationFailedReason` and `CircuitFailedReason`? What happens in the case of `FatalUpgradeError`s?

Reply:
> In my proposal (see commit above) would one bubble up ReservationFailedReason and CircuitFailedReason? 

Yes, those non-fatal errors would then be reported as behaviour event (and the listener would close).

> What happens in the case of FatalUpgradeErrors?

If I understand the implementation correctly it would add it to the pending error, which would then cause the Handler to close (thus also drop the listener) and a `SwarmEvent::ConnectionClosed` would be emitted, with
 ```rust
THandlerErr: EitherError<inbound_stop::FatalUpgradeError, outbound_hop::FatalUpgradeError>`
```

> What do you think of 2f5b8e6? Should I apply the same patterns to the other 3 UpgradeErrors.

Because of the above mentioned `SwarmEvent` I would still be in favor of reducing the  `FatalUpgradeError` to the two for the user relevant variants `IO` and `Decode`. Otherwise it looks good to me.

Reply:
Ah, returning non-fatal errors via `ProtocolsHandler::OutEvent` and fatal errors via `ProtocolsHandlerEvent::Close` makes a lot of sense. Neat, especially as it is fully typed! Thanks for the detailed suggestion @elenaf9 .

> Because of the above mentioned `SwarmEvent` I would still be in favor of reducing the `FatalUpgradeError` to the two for the user relevant variants `IO` and `Decode`. Otherwise it looks good to me.

I might be missing something. Why aggregate the fatal error types into the two (`IO` and `Decode`)? In other words, why loose the details?

I applied most of the recent suggestions to all 4 protocol files (`inbound_{hop,stop}`, `outbound_{hop,stop}`) in https://github.com/mxinden/rust-libp2p/pull/7. Shall we move the discussion there?